uber/site_sections/registration.py

```diff
@@ -801,6 +801,54 @@ def review(self, session):
                                     .order_by(Attendee.full_name).all()}
 
     @site_mappable
+    def promo(self, session, message='', **params):
+        pc = session.promo_code(params)
+        now = datetime.now(UTC)
+        today = "%d-%02d-%02d" % (now.date().year, now.date().month, now.date().day)
+        max_day = "%d-%02d-%02d" % (c.EPOCH.date().year, c.EPOCH.date().month, c.EPOCH.date().day)
+        if 'price' in params:
+            if pc.expiration_date == '':
+                pass
+            else:
+                year, month, day = pc.expiration_date.split("-")
+                pc.expiration_date = datetime(int(year), int(month), int(day)).replace(tzinfo=UTC)
+            if pc.price == '':
+                message = "Non-Negative Discounted Price is Required"
+            if pc.uses == '' or int(pc.uses) <= 0:
+                pc.uses = -1
+            if params['code'] == '':
+                pc.code = ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(6))
+            if session.query(PromoCode).filter(PromoCode.code == pc.code).first() != None:
+                message = "Promo Code Already In Use. Please Enter Another."
+
+            if message == '':
+                session.add(pc)
+                session.commit()
+                message = 'Promo Code ' + pc.code + " Generated."
```

Comment:
Generally what we do with these pages if throw the user back to the appropriate `index` page with this message using HTTPRedirect. That way, the form only reloads if there's a problem. If you anticipate that the admin will want to enter lots of codes in a row, this is fine, but you'll actually still want to use an HTTPRedirect for reasons I'll describe in the form template.